signac/contrib/migration/__init__.py

```diff
@@ -5,102 +5,135 @@
-        return version.parse(project._config["schema_version"])
-
-    if get_config_schema_version() > schema_version:
+    current_schema_version = _get_config_schema_version(root_directory, SCHEMA_VERSION)
+    if current_schema_version > schema_version:
         # Project config schema version is newer and therefore not supported.
         raise RuntimeError(
             "The signac schema version used by this project is {}, but signac {} "
             "only supports up to schema version {}. Try updating signac.".format(
-                get_config_schema_version(), __version__, SCHEMA_VERSION
+                current_schema_version, __version__, SCHEMA_VERSION
             )
         )
 
-    while get_config_schema_version() < schema_version:
-        for (origin, destination), migration in MIGRATIONS.items():
-            if version.parse(origin) == get_config_schema_version():
+    guess = current_schema_version
+    while _get_config_schema_version(root_directory, guess) < schema_version:
+        for (origin, destination), migration in _MIGRATIONS.items():
+            if version.parse(origin) == _get_config_schema_version(
+                root_directory, guess
+            ):
                 yield (origin, destination), migration
+                guess = destination
                 break
         else:
             raise RuntimeError(
                 "The signac schema version used by this project is {}, but signac {} "
                 "uses schema version {} and does not know how to migrate.".format(
-                    get_config_schema_version(), __version__, schema_version
+                    _get_config_schema_version(root_directory, guess),
+                    __version__,
+                    schema_version,
                 )
             )
 
 
-def apply_migrations(project):
+def apply_migrations(root_directory):
     """Apply migrations to a project."""
-    with _lock_for_migration(project):
-        for (origin, destination), migrate in _collect_migrations(project):
-            try:
-                print(
-                    f"Applying migration for version {origin} to {destination}... ",
-                    end="",
-                    file=sys.stderr,
-                )
-                migrate(project)
-            except Exception as e:
-                raise RuntimeError(f"Failed to apply migration {destination}.") from e
-            else:
-                _update_project_config(project, schema_version=destination)
-                print("OK", file=sys.stderr)
-                yield origin, destination
+    from ..project import Project
+
+    if isinstance(root_directory, Project):
+        warnings.warn(
+            "Migrations should be applied to a directory containing a signac project, "
+            "not a project object.",
+            FutureWarning,
+        )
+        root_directory = root_directory.root_directory()
+    lock = FileLock(os.path.join(root_directory, FN_MIGRATION_LOCKFILE))
```

Comment:
Should this go in the `try`? (move down one line?)

Reply:
I don't know the ways in which `FileLock` can fail, and I don't think this `try-finally` was designed to handle that failure specifically, but there's no harm in moving it down and allowing the lock file cleanup to happen just in case `FileLock` fails.